### ISSUE TITLE
Support for compiling CSS modules

### ISSUE THREAD

**Issue by devongovett** [parcel-bundler/parcel-css#6]
https://github.com/css-modules/css-modules

**Comment by devongovett** [comment]
May not implement `@value`. It seems like CSS variables have superseded the need for this, and it looks hard to parse since it effects every single property...

**Comment by devongovett** [comment]
`:local` and `:global` without function syntax seems like a bad idea... It's not very clear. i.e. `:global .foo .bar :local .baz` vs `.global(.foo .bar) :local(.baz)`. Most public code I could find used the function syntax, so I'm not going to implement the non-function syntax the first version and we'll see how far we can get with that.

`:import` and `:export` are low level features from ICSS, and I couldn't find any public code that used them aside from some integration tests for CSS modules.

**Comment by mohdovais** [comment]
Amazing work, congrats on launch! I see CSS modules work is under progress. Any chance providing `generateScopedName` for custom classnames similar to css-modules? I understand a callback will decrease the performance, however I use it to generate smaller names in production builds:

1. reduce css file size
2. reduce JS file size as classnames are referenced in it
3. prefix classnames with module names (SPA built with individual modules build and deployed separately)

**Comment by devongovett** [comment]
What level of logic do you need? Do you need a full callback that can do anything, or would some kind of templating syntax work for your needs? i.e. `"prefix-[name]-[hash]"`

**Comment by tanhauhau** [comment]
supporting `:local` and `:global` is still useful, especially when using with `sass` or nesting selector:

```css
:local {
   .local-1 .local-2 {
      :global {
          .global-1 { 
              color: green;
          }
      }
   }
}
```

which is normalised to

```css
:local .local-1 .local-2 :global .global-1 {
   color: green;
}
```

**Comment by tanhauhau** [comment]
would it support different `mode` like the `css-loaders` https://webpack.js.org/loaders/css-loader/#mode

reasoning: right now, once set the `css-modules` flag to be true, all selectors will be added hash, even without specifying `:local()`, and `:local()` does not work without turning `css-modules` flag on.

**Comment by devongovett** [comment]
> supporting :local and :global is still useful, especially when using with sass or nesting selector:

hmmm, interesting. I still feel this is a strange syntax though. It works differently from all other css selectors, in that it affects how components to the right are interpreted, and `:local`/`:global` do not represent elements at all. This has implications for parsing, which then requires state outside of recursion, raises questions about how nested selectors (e.g. `.foo:not(:global .bar) .baz` work, and how combinators work (e.g. `.foo :global > .bar`). With standard selectors, the components are treated independently and each one represents an element in the DOM tree. I am hesitant to support it, but if there is a very compelling use case I'm open to attempting it. I'd just want to be sure that how it should work is fully understood first.

> would it support different mode like the css-loaders

do you have a use case in mind for a css module that is global by default? isn't the point of css modules to make the classes locally scoped?

**Comment by gejgalis** [comment]
We use nested :global syntax like the following and it looks it doesn't work yet:
```scss
.foo {
  :global(.bar) {
     color:red;
  }
}
```
I expect selector: `.foo_XXX .bar { color: red }`

**Comment by devongovett** [comment]
That should work, but you need to write the `&` nesting selector and enable the nesting draft syntax (see readme). Example: https://parcel-css.vercel.app/#%7B%22minify%22%3Afalse%2C%22nesting%22%3Atrue%2C%22customMedia%22%3Afalse%2C%22cssModules%22%3Atrue%2C%22analyzeDependencies%22%3Afalse%2C%22targets%22%3A%7B%22chrome%22%3A6225920%7D%2C%22source%22%3A%22.foo%20%7B%5Cn%20%20%26%20%3Aglobal(.bar)%20%7B%5Cn%20%20%20%20%20color%3Ared%3B%5Cn%20%20%7D%5Cn%7D%5Cn%22%2C%22unusedSymbols%22%3A%5B%5D%7D

**Comment by gejgalis** [comment]
Wow @devongovett ! Thank you for that hint! I missed that part in README and I was trying to test our real use case to figure out if we can migrate to your superspeed compiler!

**Comment by dfreeman** [comment]
> May not implement @value. It seems like CSS variables have superseded the need for this, and it looks hard to parse since it effects every single property...

The "import" form like `@value moss-green from '@my-company/design-system/colors';` provides a nice mechanism for exposing shared constants (spacing increments, named colors, etc) in a way that you don't have to worry about the potential for global collisions like you would with CSS variables.

100% agreed that `@value` makes parsing hairy, and supporting the import form would also force resolution of those import paths during processing, so I understand if it's something that never gets supported here, but it does feel like a loss relative to the baseline CSS Modules feature set 🙂

**Comment by devongovett** [comment]
I have been wondering about a way to use CSS vars for the same thing. Maybe a way to enable hashing var names, and referencing/importing them from a different file, but still with the `var()` syntax so parsing is easier/closer to standard. Vars are also not global and scoped to an element which is nice.

**Comment by dfreeman** [comment]
@devongovett Reference/import-able hashed var names seem like they'd be a great way to fill the gap while building on top of platform-native features 👍 

I think something along those lines would unlock the ability for the teams I work with to consider migrating to `@parcel/css` from the PostCSS-based pipeline we use for most projects today. My Rust skills are mediocre at best, but to the degree I could help something like this land I'd be happy to lend a hand 🙂

**Comment by devongovett** [comment]
@dfreeman fyi, I implemented this syntax here: #183.

**Comment by TweededBadger** [comment]
@devongovett you may want to add to the docs what CSS Modules features are not supported (@value for example), we chased our tails for a few tails because of this.

**Comment by clarfonthey** [comment]
I think that `@value` is still valuable in cases that don't support variables, like `@media` queries. Could also implement the alternative [environment variables](https://drafts.csswg.org/css-env-1/#safe-area-insets) draft for this purpose.
